File: 2019/src/main/scala/Day4.scala
object Day4 {
  def part1(): Unit = {
    var res = 0
    for (i <- 168630 to 718098) {
      var low = 0
      var j = 0
      var in = i.toString.split("")
      var fail = false
      while (!fail && j < in.length) {
        val x = in(j).toInt
        if (x < low) {
          fail = true
        }
        low = x
        j += 1
      }
      if (!fail && in.distinct.length != in.length) res += 1
    }
    println(res)
  }

  def part2(): Unit = {
    var res = 0
    for (i <- 168630 to 718098) {
      var low = 0
      var j = 0
      var in = i.toString.split("").map(_.toInt)
      var fail = false
      var oneDouble = false
      while (!fail && j < in.length) {
        val x = in(j)
        if (x < low) {
          fail = true
        }
        else if (!oneDouble && j + 1 < in.length && x == in(j + 1) && in.count(_ == x) == 2) oneDouble = true
        low = x
        j += 1
      }
      if (!fail && oneDouble) res += 1
    }
    println(res)
  }

  def apply() = {
    println("SOLUTION DAY 4")
    println("Running part 1")
    //part1()
    println("Running part 2")
    part2()
  }
}

File: 2022/day4.py
with open("data/data4", "r") as f:
  lines = f.readlines()

def part1():
  res = 0
  for l in lines:
    [e1,e2] = l.strip().split(",")
    [l1,u1] = [int(x) for x in e1.split("-")]
    [l2,u2] = [int(x) for x in e2.split("-")]
    if l1 >= l2 and u1 <= u2:
      res += 1
    elif l2 >= l1 and u2 <= u1:
      res += 1
  print(res)

def part2():
  res = 0
  for l in lines:
    [e1,e2] = l.strip().split(",")
    [l1,u1] = [int(x) for x in e1.split("-")]
    [l2,u2] = [int(x) for x in e2.split("-")]
    if l1 >= l2 and l1 <= u2:
      res += 1
    elif u1 >= l2 and u1 <= u2:
      res += 1
    elif l2 >= l1 and l2 <= u1:
      res += 1
    elif u2 >= l1 and u2 <= u1:
      res += 1
  print(res)

part1()
part2()